src/context/WalletProvider/NativeWallet/components/NativeTestPhrase.tsx

```diff
@@ -1,225 +1,185 @@
-  )
 
   const { vault } = location.state
 
-  const shuffleMnemonic = useCallback(async () => {
-    if (testCount.current >= TEST_COUNT_REQUIRED) return
+  const generateTestState = useCallback(async () => {
     try {
       const mnemonic = await vault.unwrap().get('#mnemonic')
       const words = mnemonic.split(' ')
-      let randomWords = uniq(bip39.generateMnemonic(256).split(' ')) as string[]
 
-      const targetWordIndex = shuffledNumbers[testCount.current]
-      const targetWord = words[targetWordIndex]
-      randomWords = randomWords.filter(x => x !== targetWord).slice(0, 3)
-      randomWords.push(targetWord)
-      randomWords = shuffle(randomWords)
-      const correctAnswerIndex = randomWords.indexOf(targetWord)
-      // Should never happen because we literally just added the word to the array
-      if (correctAnswerIndex === -1) throw Error("Can't find index of current word in randomWords")
+      const targetIndices = getRandomIndicesIndexes(words.length, TEST_COUNT_REQUIRED)
+      const options = targetIndices.map(idx => {
+        const correct = words[idx]
+        const distractors = getRandomWords(bip39.wordlists.english, correct, 2)
+        return uniq([correct, ...distractors]).sort(() => Math.random() - 0.5)
+      })
 
       setTestState(
         revocable(
           {
-            targetWordIndex,
-            randomWords,
-            correctAnswerIndex,
+            targetIndices,
+            options,
+            selectedWords: [],
           },
           revoker.addRevoker.bind(revoker),
         ),
       )
     } catch (e) {
       setError('walletProvider.shapeShift.create.error')
     }
-  }, [setTestState, shuffledNumbers, vault, revoker, testCount])
+  }, [vault, revoker])
 
   useEffect(() => {
-    if (!isInitiallyShuffled.current) {
-      shuffleMnemonic()
-      isInitiallyShuffled.current = true
-    }
-  }, [shuffleMnemonic])
+    generateTestState()
+  }, [generateTestState])
 
   const handleBackupComplete = useCallback(() => {
     vault.seal()
     navigate('/native/password', { state: { vault }, replace: true })
     setTimeout(() => revoker.revoke(), 250)
   }, [navigate, revoker, vault])
 
-  const handleClick = (index: number) => {
-    if (index === testState?.correctAnswerIndex) {
-      testCount.current++
-
-      if (testCount.current >= TEST_COUNT_REQUIRED) {
-        handleBackupComplete()
-      } else {
-        shuffleMnemonic()
-      }
-    } else {
-      shuffleMnemonic()
-    }
+  const handleSelect = (line: number, index: number) => {
+    if (!testState) return
+
+    setTestState(prev => {
+      if (!prev) return prev
+      const next = { ...prev }
+      next.selectedWords[line] = testState.options[line][index]
+      return revocable(next, revoker.addRevoker.bind(revoker))
+    })
   }
 
+  const { data: isCorrect } = useQuery({
+    queryKey: ['isCorrect', testState?.selectedWords],
+    queryFn: async () => {
+      if (!testState) return false
```

Comment:
nitpick: can do `testState ? queryFn : skipToken` pattern instead of old `enabled` pattern, and avoid this